Added migration guide for Link SDK (#1626)
codatio/codat-docs: docs/auth-flow/migration-guide.md

---
title: "Migration guide"
description: "Learn how to migrate from a custom link flow to our embedded Link component"
---

import CodeExamples from "./_partial-auth-flow-examples.md";

## Overview

This guide helps you migrate from a custom link flow to our embedded **Link** component. The embedded Link offers a more streamlined, secure, and maintainable way to connect your customers’ financial software.

## Prerequisites

> You need a JavaScript application to render the component. Link works with all major JavaScript frameworks—including React—and also with vanilla JavaScript. TypeScript is supported and recommended.

Reviewer: ```suggestion
You need a JavaScript application to render the Link component. Our component works with vanilla JavaScript and all major JavaScript frameworks. TypeScript is supported and recommended.
```

:::warning Do not use Link inside an iframe. It won’t work due to [CORS](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) restrictions.
:::

## Your current implementation

### User experience

Your current implementation likely starts with a consent page, confirming that your SMB customers are happy to share their financial data with you before they interact with Codat.
This often includes information on the purpose, platform data accessed, and your retention policy. 

Codat's Link component optionally supports all of this and more, so make sure to review your current implementation to understand what changes you'd like to make.

### Custom link workflow

There are two common approaches for custom link flows, depending on your use case:

- **Create a company with a platformKey:**  
  You pass the company name and the financial software’s `platformKey` in one API call. Best suited for users connecting to a single financial system.

- **Create company and connection separately:**  

>   You create the company first, then create a connection. This supports multiple financial systems.

Reviewer: ```suggestion
  You create the company first, then create a connection. This approach is typically followed when connecting several platforms for one customer e.g. Accounting Platform and Banking.
```

![Custom Link workflow](/img/auth-flow/migration-guide/custom-link-flow.png)

## The Link approach

The embedded Link component replaces your custom implementation while preserving the user experience elements you've already built. Instead of managing connection logic, API calls, and UI components in your frontend:

> 1. Your backend requests an access token from Codat’s API.

Reviewer 2: This is the functionality we don't currently support right?

Reviewer: Yes and I want @andy-white-hat to validate this approach is not a security concern.

Author: Correct.

> 2. You pass the access token and company name to the Link component.

Reviewer 2: @jbyrnecodat can maybe confirm this, I'm not sure we can start the Link component with just a company name? We need a company ID

Reviewer: @jbyrnecodat @pmckinney-codat - are we happy this is a change we **can** and are _willing_ to make?

Feels like a relatively significant departure from current implementations and behaviour.

3. The user completes the connection flow in the embedded UI.

> Link supports both of your current approaches—whether you're using the platformKey flow for single connections or the two-step approach for multiple software connections.

Reviewer 3: ```suggestion
Link supports both of your current approaches — whether you're using the platform key flow for single connections or the two-step approach for multiple software connections.
```

![Embedded Link workflow](/img/auth-flow/migration-guide/link-flow.png)

### Retain your current user experience

Your existing consent page, data sharing policies, and user onboarding flow don't need to change. Link is highly customizable, allowing you to:

- **Preserve your consent flow:** Use custom consent pages with your existing messaging about data purpose and retention
- **Maintain your branding:** Control colors, logos, and styling to match your application
- **Choose your layout:** Display Link in modal or non-modal views based on your current UI patterns
- **Control the journey:** Decide when and how users interact with the connection flow

> This means you can migrate to Link while keeping the user experience your customers are already familiar with.

Reviewer: ```suggestion
This means you can migrate your implementation to leverage the features of our Link product, whilst maintaining your existing customer experience and consent journey.
```

Learn how to [customize your auth flow](/auth-flow/customize/sdk-customize-code)

## Migration steps

### Step 1: Set up CORS

[Cross-origin resource sharing](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) (CORS) settings are required for access tokens to work. Configure CORS settings on your Codat instance:

- Use [Set CORS](/platform-api#/operations/set-cors-settings) settings to register allowed domains
- Use [Get CORS](/platform-api#/operations/get-cors-settings) settings to view your current configuration

> ### Step 2: Implement access token retrieval

Reviewer: @pmckinney-codat @jbyrnecodat I don't like this wording and is it correct?

1. Create a backend endpoint to proxy Codat's API
2. In that endpoint, call `GET /accessToken` to retrieve an access token for the Link component
3. Return the `accessToken` in the response

### Step 3: Embed Link component

:::tip Install the npm package

> Take advantage of our [npm package](https://www.npmjs.com/package/@codat/sdk-link-types) so you don't have to manually import and maintain type definitions. You will benefit from it the most if you are using Typescript, so our examples are prepared with that assumption.

Reviewer: ```suggestion
Codat provides a `types` package on [npm](https://www.npmjs.com/package/@codat/sdk-link-types) that contains type definitions for our Link component primitives. We recommend using this package as it simplifies your implementation.
```

`npm install @codat/sdk-link-types`

:::

<CodeExamples />

Configure Link to match your current implementation:

> - **Choose your layout:** Configure Link to display in modal or non-modal views to match your current UI

Reviewer 2: ```suggestion
- **[Choose your layout:](/auth-flow/customize/sdk-customize-code#properties)** Configure Link to display in modal or non-modal views to match your current UI
```
